### PR TITLE
Opt slice fix

### LINKED ISSUE #1235: dask.array regression involving arithmetic with xarray objects
This code worked with dask v0.7.6, but failed when I updated to v0.9.0:
```
In [1]: import numpy as np

In [2]: import xarray as xr

In [3]: import dask.array as da

In [4]: x = xr.core.indexing.LazilyIndexedArray(np.zeros((3, 3)))

In [5]: y = da.from_array(x, (2, 2))

In [6]: (y[:2, :2] * y[:2, :2]).compute()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-82666a5a7d29> in <module>()
----> 1 (y[:2, :2] * y[:2, :2]).compute()

/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/base.py in compute(self, **kwargs)
     35
     36     def compute(self, **kwargs):
---> 37         return compute(self, **kwargs)[0]
     38
     39     @classmethod

/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/base.py in compute(*args, **kwargs)
    108                 for opt, val in groups.items()])
    109     keys = [var._keys() for var in variables]
--> 110     results = get(dsk, keys, **kwargs)
    111
    112     results_iter = iter(results)

/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/threaded.py in get(dsk, result, cache, num_workers, **kwargs)
     55     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     56                         cache=cache, queue=queue, get_id=_thread_get_id,
---> 57                         **kwargs)
     58
     59     return results

/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/async.py in get_async(apply_async, num_workers, dsk, result, cache, queue, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, **kwargs)
    486                 _execute_task(task, data)  # Re-execute locally
    487             else:
--> 488                 raise(remote_exception(res, tb))
    489         state['cache'][key] = res
    490         finish_task(dsk, key, state, results, keyorder.get)

TypeError: unsupported operand type(s) for *: 'LazilyIndexedArray' and 'LazilyIndexedArray'

Traceback
---------
  File "/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/async.py", line 267, in execute_task
    result = _execute_task(task, data)
  File "/Users/shoyer/conda/envs/xarray-dev/lib/python3.5/site-packages/dask/async.py", line 249, in _execute_task
    return func(*args2)
```

LazilyIndexedArray is an object that satisfies some of the usual duck-array API (e.g. shape, dtype, `__getitem__`) and that can be coerced into a NumPy array:
https://github.com/pydata/xarray/blob/v0.7.2/xarray/core/indexing.py#L272

I *think* it should be valid into to `da.from_array` -- this certainly worked in old versions of dask.

### PR DESCRIPTION
**Pull request by jcrist**
When slicing an unknown array-like object, ``getarray`` is used to ensure that the output of the slicing operation is a numpy array, instead of another array-like. When optimizing nested slices, we need to
ensure that if ``getarray`` is used at all in the chain, then the slicing still happens. A bug in this logic was introduced in the recent work on speeding up the optimization passes. This patch fixes this bug, and reverts the changes to the tests made that allowed it to pass.

As a separate commit, also removed unused imports, and fixed several indenting pep8 things (mostly due to indent lining up with block, making if blocks harder to read).

Fixes #1235.

### CODE REVIEW

**Comment by jcrist** [comment]
cc @shoyer.

**Comment by shoyer** [comment]
Maybe worth adding something resembling a regression test? If dask.array doesn't already have tests with non-ndarray array-like objects those would be nice to have -- in all important respects, my LazilyIndexedArray works like a h5py.Dataset here.

**Comment by jcrist** [comment]
The optimization tests (after reverting them back) are sufficient to catch this, but I also added another test slicing from a `np.matrix`. `h5py.Dataset`s return `np.ndarray`s when sliced, while `np.matrix`s don't.

**Comment by mrocklin** [comment]
Downgrading the optimization tests was my fault.  I didn't realize that
they were catching special behavior.  Perhaps there is a more explicit test
that could be added.

**Comment by jcrist** [comment]
Is the `np.matrix` test here good for that (it checks that nested slices still return np.ndarray)? Or were you suggesting another optimization test to look at the graph? Perhaps just a comment in that test explaining why the test exists?

**Comment by mrocklin** [comment]
Fine with me

File: dask/array/tests/test_array_core.py [hunk 1]
```diff
@@ -1296,6 +1293,17 @@ def test_from_array_with_lock():
     assert_eq(e + f, x + x)
 
 
+def test_from_array_slicing_results_in_ndarray():
```
Comment (shoyer) [inline dask/array/tests/test_array_core.py]: This looks good, thanks!

**Comment by jcrist** [comment]
Ok. Merging.

**Pull request merged**